### ISSUE TITLE
Website deployment sometimes hangs if retry is triggered

### ISSUE THREAD

**Issue by droyad** [OctopusDeploy/Issues#2824]
Reported in this support post: http://help.octopusdeploy.com/discussions/problems/47828-applicatiohostconfig-retry-issue

When multiple web deployments are run simultaneously on the same server, they may clash writing to the IIS Metabase. This causes the deployment script to sleep and then retry several times. Occasionally the script will never resume from the sleep and the powershell session is running but inactive. This occurs about 5-50% of test runs.

This can only occur if the deployments have `OctopusBypassDeploymentMutex` set to true, bypassing the protections. It may be problematic making simultaneous writes to IIS metabase in any case, so we recommend not enabling this feature if that is the case.

I am unsure whether this is related to the IIS metabase writing or any other operations in the script. It could be just this script as it has a high probability of triggering sleep-retry

It hangs on this line: https://github.com/OctopusDeploy/Calamari/blob/master/source/Calamari/Scripts/Octopus.Features.IISWebSite_BeforePostDeploy.ps1#L72. The `Retrying...` line is never printed.

When the debugger is attached, there is nothing obvious other than two threads waiting on a mutex. Unsure if it is the same mutex or if they are deadlocking each other.

Moderate CPU/Memory usage seems to increase the occurrence.

**Comment by droyad** [comment]
To reproduce in Octopus, create 3 projects each deploying a website. Use the exact same steps to ensure their timing coincides. Deploy all 3 to different environments as fast as possible (using `Octo.exe` to trigger them immediately after each other).

I have reduced the problem to the three scripts in [this Zip](https://github.com/OctopusDeploy/Issues/files/523523/Freeze.zip). Run `Test.ps1`, which will execute multiple instances of the deployment scripts in new windows over and over again. One of the windows will eventually hang (1-5 mins).

Code has been removed from the scripts in an attempt to narrow down the problem.
